chronos/asyncfutures2.nim

```diff
@@ -784,21 +799,83 @@ proc cancelAndWait*(fut: FutureBase): Future[void] =
   ##
   ## If ``fut`` is already finished (completed, failed or cancelled) result
   ## Future[void] object will be returned complete.
-  var retFuture = newFuture[void]("chronos.cancelAndWait(T)")
-  proc continuation(udata: pointer) =
-    if not(retFuture.finished()):
-      retFuture.complete()
-  proc cancellation(udata: pointer) =
-    if not(fut.finished()):
-      fut.removeCallback(continuation)
+  let retFuture = newFuture[void]("chronos.cancelAndWait(FutureBase)",
+                                  {FutureFlag.OwnCancelSchedule})
+
+  proc continuation(udata: pointer) {.gcsafe.} =
+    retFuture.complete()
+
+  proc cancellation(udata: pointer) {.gcsafe.} =
+    # We are not going to change the state of `retFuture` to cancelled, so we
+    # will prevent the entire sequence of Futures from being cancelled one more
+    # time.
+    discard
+
   if fut.finished():
     retFuture.complete()
   else:
     fut.addCallback(continuation)
     retFuture.cancelCallback = cancellation
     # Initiate cancellation process.
     fut.cancel()
-  return retFuture
+  retFuture
+
+proc checkedCancelAndWait*(fut: FutureBase): Future[bool] =
+  ## Initiate cancellation process for Future ``fut`` and wait until ``fut`` is
+  ## done e.g. changes its state (become completed, failed or cancelled).
```

Comment:
```suggestion
  ## finished e.g. changes its state (become completed, failed or cancelled).
```

Reply:
Initially it was supposed that `cancel()` should return boolean flag which indicates:
`true` if cancellation was successful - e.g. last Future in sequence was moved from `Pending` to `Cancelled`, all future callbacks got scheduled.
`false` if cancellation was unsuccessfull, e.g. we encounter any Future in sequence which are not `Pending`. It means that this sequence of Futures are moving towards finish and we can't cancel it.

But somehow this flag was discarded and not used, so i expect there present a lot of source code which do not respect boolean result and do not expect `cancel(): boolean`. So `checkCancelAndWait` it is same as `cancelAndWait` but it returns flag which indicates is future was actually cancelled or not.

Reply:
what does the boolean mean though? It feels like it's too weak to be useful - an async operation is typically made up of many sub-operations and when we start cancelling things, some of these may have completed and some may be cancelled - I'm not sure a boolean alone can capture this complexity in a useful way and this might be why everyone discards it.

Basically, there are three options: "it was cancelled", "it finished" and "don't know/partial result" and the boolean doesn't represent this well enough to be useful practically.

Reply:
Cancellation could be compared to "message delivery", so this boolean flag indicates whether message was delivered to the bottom layer of target `Future` or it was not delivered.

> what does the boolean mean though? It feels like it's too weak to be useful - an async operation is typically made up of many sub-operations and when we start cancelling things, some of these may have completed and some may be cancelled - I'm not sure a boolean alone can capture this complexity in a useful way and this might be why everyone discards it.

chronos tracks dependent `Futures` and cancellation using this tracking mechanism to deliver cancellation message to the bottom layer. So if bottom layer or any other Future before bottom layer will have `status != Pending` it will stop at this moment and returns `false`. But if the bottom layer Future has `status == Pending`, it will become `Cancelled`. And in this case `cancel()` will return `true`.